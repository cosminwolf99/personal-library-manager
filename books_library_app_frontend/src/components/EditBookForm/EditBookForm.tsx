import React, { useState } from "react";
import { useFormik } from "formik";
import { KeyedMutator } from "swr";
import {
  TextField,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
} from "@mui/material";
import { Book } from "../../types/book-types";
import { updateBook } from "../../api/books/updateBook";
import validationSchema from "../../utils";

interface EditBookFormProps {
  book: Book;
  mutate: KeyedMutator<Book[]>;
}

const EditBookForm: React.FC<EditBookFormProps> = ({ book, mutate }) => {
  const [open, setOpen] = useState(false);

  const handleClickOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  const formik = useFormik({
    initialValues: {
      title: book.title,
      author: book.author,
      genre: book.genre,
      description: book.description,
    },
    validationSchema: validationSchema,
    onSubmit: async (values) => {
      await updateBook({ ...book, ...values });
      mutate();
      handleClose();
    },
  });

  const hasChanges = Object.keys(formik.values).some(
    (key) =>
      formik.values[key as keyof typeof formik.values] !==
      book[key as keyof typeof book]
  );

  return (
    <>
      <Button onClick={handleClickOpen} color="primary">
        Edit
      </Button>
      <Dialog open={open} onClose={handleClose}>
        <DialogTitle>Edit Book</DialogTitle>
        <DialogContent>
          <form onSubmit={formik.handleSubmit}>
            <TextField
              fullWidth
              id="title"
              name="title"
              label="Title"
              value={formik.values.title}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              error={formik.touched.title && Boolean(formik.errors.title)}
              helperText={formik.touched.title && formik.errors.title}
              margin="normal"
            />
            <TextField
              fullWidth
              id="author"
              name="author"
              label="Author"
              value={formik.values.author}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              error={formik.touched.author && Boolean(formik.errors.author)}
              helperText={formik.touched.author && formik.errors.author}
              margin="normal"
            />
            <TextField
              fullWidth
              id="genre"
              name="genre"
              label="Genre"
              value={formik.values.genre}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              error={formik.touched.genre && Boolean(formik.errors.genre)}
              helperText={formik.touched.genre && formik.errors.genre}
              margin="normal"
            />
            <TextField
              fullWidth
              id="description"
              name="description"
              label="Description"
              value={formik.values.description}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              error={
                formik.touched.description && Boolean(formik.errors.description)
              }
              helperText={
                formik.touched.description && formik.errors.description
              }
              margin="normal"
              multiline
              rows={4}
            />
          </form>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose} color="primary">
            Cancel
          </Button>
          <Button
            onClick={() => formik.handleSubmit()}
            color="primary"
            disabled={!hasChanges || !formik.isValid || formik.isSubmitting}
          >
            Save
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default EditBookForm;
