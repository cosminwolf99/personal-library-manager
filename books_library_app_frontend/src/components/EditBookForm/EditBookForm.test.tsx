/* eslint-disable testing-library/prefer-screen-queries */
import { render, fireEvent } from "@testing-library/react";
import EditBookForm from "./EditBookForm";
import { Book } from "../../types/book-types";

jest.mock("../../api/books/updateBook", () => ({
  updateBook: jest.fn(),
}));

jest.mock("swr", () => ({
  mutate: jest.fn(),
}));

const mockBook: Book = {
  id: 1,
  title: "Test Book",
  author: "Test Author",
  genre: "Test Genre",
  description: "Test Description",
};

describe("EditBookForm", () => {
  it("renders correctly and matches snapshot", () => {
    const { container } = render(
      <EditBookForm book={mockBook} mutate={jest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the edit dialog when the Edit button is clicked", () => {
    const { getByText, getByLabelText } = render(
      <EditBookForm book={mockBook} mutate={jest.fn()} />
    );

    // Ensure the "Edit" button is rendered
    const editButton = getByText(/edit/i);
    expect(editButton).toBeInTheDocument();

    // Simulate clicking the edit button
    fireEvent.click(editButton);

    // Check that the dialog with the form fields is rendered
    expect(getByLabelText(/Title/i)).toBeInTheDocument();
    expect(getByLabelText(/Author/i)).toBeInTheDocument();
    expect(getByLabelText(/Genre/i)).toBeInTheDocument();
    expect(getByLabelText(/Description/i)).toBeInTheDocument();
  });
});
