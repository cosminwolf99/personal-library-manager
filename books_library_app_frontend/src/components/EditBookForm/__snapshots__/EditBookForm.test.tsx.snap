// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditBookForm renders correctly and matches snapshot 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1apkmlw-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Edit
  </button>
</div>
`;
